core/src/main/java/org/apache/iceberg/rest/http/IcebergHttpResponse.java

```diff
@@ -0,0 +1,111 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.iceberg.rest.http;
+
+import com.fasterxml.jackson.annotation.JsonCreator;
+import com.fasterxml.jackson.annotation.JsonInclude;
+import com.fasterxml.jackson.annotation.JsonProperty;
+
+/**
+ * All responses for the first version of the Iceberg REST API will be JSON objects, a top-level envelope object with two fields,
+ * `error` and `data`, which are themselves JSON objects.
+ *
+ * data: represents the JSON encoded response for a successful path or for a valid / expected failure. JSON representation of various response types
+ * error: a standardized obect on error, containing the error code, message, type, and  additional metadata (an optional JSON object of metadata) as defined below.
+ *
+ * All responses for the REST catalog should be wrapped this way, vs using primitives. For example, for a ListTableResponse, listing tables under a namespace "accounting",
+ * we'd get a JSON object back like the following:
+ *
+ * { "data": { "identifiers": [ "accounting.tax", "accounting.currency_conversions"] }, "error", {} }
+ *
+ * If the namesapce `accounting` didn't request, the response from that call would have a body like the following,
+ * where the `code` 40401 is a two-part identifier:
+ *    - HTTP response code: 404
+ *    - Two digit internal application defined error code for further detail: 01 - Namespace not found.
+ *
+ * { "data": {}, "error": { "message": "Failed to list tables. The Namespace 'accounting' does not exist", "type": "NamespaceNotFoundException", "code": 40401 }
+ *
+ * We could also embed the HTTP response code plainly by itself, without internally documented codes, as a separate field. I have found having a documented list of internal
+ * error codes to be very helpful previously, but am open to discussion on this.
```

Comment:
Could it be clearer if the HTTP error code were separate from the internal error code? This would allow free form internal code without having to be confined to 100 values under an HTTP code prefix, and at the same time maintain reference to the more generic HTTP set of codes. It would also make it easier to extract HTTP error codes for processing.

Reply:
So since we have `type`, which can be used to enumerate all of the possible sub fields, we've moved in the direction of just using the 3-digit HTTP response code in the body.

This file isn't part of the spec at all, as mentioned below,  so I forgot to update it. I will remove it for now actually as it was just to show somebody something.

Do you input on using dedicated `code` (perhaps not constrained to HTTP code space as you mentioned) instead of just using the standardized value in `type` (which is a string)?

If we used `code`, the lookup table in my experience brings you to `type`. But I have previously used the way you described, with numeric error codes that we define a lookup table for that are independent. Do you have feedback on moving to just `type`?

Reply:
IMHO, `type` is preferable to numeric `code`, but it might be harder to define strictly and there's risk that `type` names deviate from actual exception class names. Plus fixing class names in the API spec might affect non-java implementations in a negative way.

In Nessie, as an example, we have an enum of non-numeric error codes. The API returns enum names, but they can be mapped to java exceptions upon parsing. Python client has its own mapping too. cf. [ErrorCode.java](https://github.com/projectnessie/nessie/blob/main/model/src/main/java/org/projectnessie/error/ErrorCode.java)

Reply:
Excluding this class from the spec and using only OpenAPI yaml SGTM :+1: 

Re: `type` values, I personally tend to think that if the field itself is part of the spec, the spec would have to define how to interpret it. I am not sure keeping it "opaque" is going to be that much useful from the spec's perspective. 

Is this spec intended to define just a common part of the protocol between a Catalog impl. on the Iceberg side and the HTTP server that offers the REST API? Are the client and server parts supposed to be interoperable or are they assumed to work knowing the exact nature and version of the other side?

Reply:
> Re: type values, I personally tend to think that if the field itself is part of the spec, the spec would have to define how to interpret it. I am not sure keeping it "opaque" is going to be that much useful from the spec's perspective.

I agree. Since we're using code to just be the HTTP status response code, I would say that `type` should be a well defined set we can look up.

We might also choose to allow people to use others that will get ignored by the client but are logged or something, but for now I would say let's plan for it to be a well defined set of values, but not necessarily attempt to enumerate all of them yet since we don't have about half of the basic Catalog AP in here yet. It's currently mostly just around namespaces and some of the table related ones will be handled next.

Sound good?